syntax = "proto3";

package suoke.benchmark.v1;

// Benchmark服务定义
service BenchmarkService {
  // 运行基准测试
  rpc RunBenchmark (RunBenchmarkRequest) returns (RunBenchmarkResponse);
  
  // 获取基准测试结果
  rpc GetBenchmarkResult (GetBenchmarkResultRequest) returns (GetBenchmarkResultResponse);
  
  // 获取支持的基准测试任务
  rpc ListBenchmarks (ListBenchmarksRequest) returns (ListBenchmarksResponse);
  
  // 比较两个模型版本
  rpc CompareBenchmarks (CompareBenchmarksRequest) returns (CompareBenchmarksResponse);
  
  // 导出评测报告
  rpc ExportReport (ExportReportRequest) returns (ExportReportResponse);
  
  // 流式监控评测进度
  rpc MonitorBenchmark (MonitorBenchmarkRequest) returns (stream BenchmarkProgress);
}

// 运行基准测试请求
message RunBenchmarkRequest {
  // 基准测试配置
  string benchmark_id = 1;  // 基准测试ID
  string model_id = 2;      // 模型ID
  string model_version = 3; // 模型版本
  BenchmarkTask task = 4;   // 任务类型
  map<string, string> parameters = 5;  // 附加参数
}

// 运行基准测试响应
message RunBenchmarkResponse {
  string run_id = 1;        // 运行ID
  string status = 2;        // 运行状态
  string message = 3;       // 状态信息
}

// 获取基准测试结果请求
message GetBenchmarkResultRequest {
  string run_id = 1;        // 运行ID
  bool include_details = 2; // 是否包含详细结果
}

// 获取基准测试结果响应
message GetBenchmarkResultResponse {
  string run_id = 1;                 // 运行ID
  string benchmark_id = 2;           // 基准测试ID
  string model_id = 3;               // 模型ID
  string model_version = 4;          // 模型版本
  string status = 5;                 // 运行状态
  map<string, MetricResult> metrics = 6;  // 指标结果
  repeated SampleResult samples = 7;      // 样本结果
  BenchmarkTask task = 8;                 // 任务类型
  string created_at = 9;                  // 创建时间
  string completed_at = 10;               // 完成时间
}

// 指标结果
message MetricResult {
  string name = 1;          // 指标名称
  double value = 2;         // 指标值
  string unit = 3;          // 单位
  double threshold = 4;     // 阈值
  bool pass = 5;            // 是否通过
  string comparison = 6;    // 与基线比较
}

// 样本结果
message SampleResult {
  string sample_id = 1;           // 样本ID
  string input = 2;               // 输入
  string expected = 3;            // 期望输出
  string actual = 4;              // 实际输出
  bool correct = 5;               // 是否正确
  map<string, double> scores = 6; // 详细分数
}

// 列出基准测试请求
message ListBenchmarksRequest {
  BenchmarkTask task_filter = 1;  // 任务类型过滤
  string tag = 2;                 // 标签过滤
}

// 列出基准测试响应
message ListBenchmarksResponse {
  repeated BenchmarkInfo benchmarks = 1;  // 基准测试列表
}

// 基准测试信息
message BenchmarkInfo {
  string id = 1;                // 基准测试ID
  string name = 2;              // 名称
  string description = 3;       // 描述
  BenchmarkTask task = 4;       // 任务类型
  repeated string tags = 5;     // 标签
  uint32 sample_count = 6;      // 样本数量
  repeated string metrics = 7;  // 支持的指标
  map<string, string> parameters = 8;  // 参数说明
}

// 比较基准测试请求
message CompareBenchmarksRequest {
  string baseline_run_id = 1;   // 基线运行ID
  string compare_run_id = 2;    // 对比运行ID
}

// 比较基准测试响应
message CompareBenchmarksResponse {
  string baseline_model = 1;    // 基线模型
  string compare_model = 2;     // 对比模型
  map<string, ComparisonResult> metrics = 3;  // 指标比较
  repeated CaseComparison case_comparisons = 4;  // 案例比较
  string summary = 5;           // 比较总结
}

// 指标比较结果
message ComparisonResult {
  double baseline_value = 1;    // 基线值
  double compare_value = 2;     // 对比值
  double diff = 3;              // 差值
  double diff_percent = 4;      // 差值百分比
  bool significant = 5;         // 是否显著
}

// 案例比较
message CaseComparison {
  string sample_id = 1;         // 样本ID
  string input = 2;             // 输入
  string expected = 3;          // 期望输出
  string baseline_output = 4;   // 基线输出
  string compare_output = 5;    // 对比输出
  bool baseline_correct = 6;    // 基线是否正确
  bool compare_correct = 7;     // 对比是否正确
}

// 导出报告请求
message ExportReportRequest {
  string run_id = 1;            // 运行ID
  ReportFormat format = 2;      // 报告格式
  bool include_samples = 3;     // 是否包含样本
  repeated string metrics = 4;  // 要包含的指标
}

// 导出报告响应
message ExportReportResponse {
  string report_url = 1;        // 报告URL
  string message = 2;           // 状态信息
}

// 监控基准测试请求
message MonitorBenchmarkRequest {
  string run_id = 1;            // 运行ID
}

// 基准测试进度
message BenchmarkProgress {
  string run_id = 1;            // 运行ID
  string status = 2;            // 状态
  double progress = 3;          // 进度 (0-100)
  string current_stage = 4;     // 当前阶段
  uint32 processed_samples = 5; // 已处理样本数
  uint32 total_samples = 6;     // 总样本数
  string message = 7;           // 状态信息
  map<string, double> current_metrics = 8;  // 当前指标
}

// 报告格式枚举
enum ReportFormat {
  HTML = 0;
  PDF = 1;
  JSON = 2;
  MARKDOWN = 3;
}

// 基准测试任务类型
enum BenchmarkTask {
  UNKNOWN = 0;
  TCM_DIAGNOSIS = 1;          // 中医辨证
  TONGUE_RECOGNITION = 2;     // 舌象识别
  FACE_RECOGNITION = 3;       // 面色识别
  PULSE_RECOGNITION = 4;      // 脉象识别
  HEALTH_PLAN_GENERATION = 5; // 健康方案生成
  AGENT_COLLABORATION = 6;    // 智能体协作
  PRIVACY_VERIFICATION = 7;   // 隐私验证
  EDGE_PERFORMANCE = 8;       // 端侧性能
  DIALECT_RECOGNITION = 9;    // 方言识别
}